source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs

```diff
@@ -0,0 +1,710 @@
+
+                if (value is string && isTemplatedString)
+                {
+                    result.Append("\"");
+                }
+            }
+            else
+            {
+                result.Append("${" + unboundString + "}");
+            }
+
+            return result.ToString();
+        }
+
+        /// <summary>
+        /// return the parsed result of $when from pair context
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public override AdaptiveCardsTemplateResult VisitTemplateWhen([NotNull] AdaptiveCardsTemplateParser.TemplateWhenContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+            // when this node is visited, the children of this node is shown as below: 
+            // this node is visited only when parsing was correctly done
+            // [ '{', '$when', ':', ',', 'expression'] 
+            var result = Visit(context.templateExpression());
+            return result;
+        }
+
+        /// <summary>
+        /// Visit method for <c>array</c> grammar in <c>AdaptiveCardsTemplateParser.g4</c>
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns>AdaptiveCardsTemplateResult</returns>
+        public override AdaptiveCardsTemplateResult VisitArray([NotNull] AdaptiveCardsTemplateParser.ArrayContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.exception != null)
+            {
+                return new AdaptiveCardsTemplateResult(context.GetText());
+            }
+
+            AdaptiveCardsTemplateResult result = new AdaptiveCardsTemplateResult(context.LSB().GetText());
+            var values = context.value();
+            var arrayDelimiters = context.COMMA();
+
+            // visit each json value in json array and integrate parsed result
+            for(int i = 0; i < values.Length; i++)
+            {
+                var value = context.value(i);
+                var parsedResult = Visit(value);
+                result.Append(parsedResult);
+                // only add delimiter when parsedResult has not been dropped, and delimiter is needed
+                if (!parsedResult.HasItBeenDropped && i != values.Length - 1 && arrayDelimiters.Length > 0)
+                {
+                    result.Append(arrayDelimiters[0].GetText());
+                }
+            }
+
+            result.Append(context.RSB().GetText());
+
+            return result;
+        }
+        public static bool IsTrue(string predicate, JToken data)
+        {
+            var (value, error) = new ValueExpression(predicate).TryGetValue(data);
+            if (error == null)
+            {
+                return bool.Parse(value as string);
+            }
+            return true;
```

Comment:
so if there's an error, we return `true`?

Reply:
when predicate is evaluated to false, we want to do an action of dropping json obj, otherwise we want to keep the string we are parsing without changes.

---
In reply to: [414190125](https://github.com/microsoft/AdaptiveCards/pull/3946#discussion_r414190125) [](ancestors = 414190125)